### PR TITLE
Parallel Conv, Deconv and CNN

### LINKED ISSUE #561: TTP: ParallelCNN

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Add parallel support to CNN.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+                 strides=1,
+                 padding=0,
+                 use_bias=True,
+                 kernel_initializer=None,
+                 kernel_init_gain=1.0,
+                 bias_init_value=0.0):
+        """A parallel 2D Conv layer that's also responsible for activation and customized
+        weights initialization. An auto gain calculation might depend on the
+        activation following the conv layer. Suggest using this wrapper module
+        instead of ``nn.Conv2d`` if you really care about weight std after init.
```
Comment (hnyu) [inline alf/layers.py]: Add one sentence about the purpose/functionality of this layer

File: alf/layers.py [hunk 2]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+
+        Args:
+            in_channels (int): channels of the input image
+            out_channels (int): channels of the output image
+            kernel_size (int or tuple):
+            n (int): n indepedent ``Conv2D`` layer
```
Comment (hnyu) [inline alf/layers.py]: layer -> layers

File: alf/layers.py [hunk 3]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+            activation (torch.nn.functional):
+            strides (int or tuple):
+            padding (int or tuple):
+            use_bias (bool):
+            kernel_initializer (Callable): initializer for the conv layer kernel.
+                If None is provided a variance_scaling_initializer with gain as
```
Comment (hnyu) [inline alf/layers.py]: \``variance_scaling_initializer\``

File: alf/layers.py [hunk 4]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+            nn.init.constant_(self._conv2d.bias.data, bias_init_value)
+
+    def forward(self, img):
+        """Forward
+
+        Args:
+            img (torch.Tensor): with shape ``[B, C, H, W]``
+                                        or ``[B, n, C, H, W]``
+            where the meaning of the symbols are:
+            B: batch, C: number of channels, H: image height, W: image width
```
Comment (hnyu) [inline alf/layers.py]: The format of this argument will be broken; explain the difference between with n and without n

File: alf/layers.py [hunk 5]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+        Returns:
+            torch.Tensor with shape ``[B, n, C', H', W']``
```
Comment (hnyu) [inline alf/layers.py:434]: broken format; can use a bullet list to explain the symbols

File: alf/layers.py [hunk 6]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+
+        # merge replica and channels
+        img = img.reshape(img.shape[0], img.shape[1] * img.shape[2],
+                          *img.shape[3:])
+
+        res = self._activation(self._conv2d(img))
+
+        # reshape back: [B, n*C', H', W'] -> [B, n, C', H', W']
+        res = res.reshape(res.shape[0], self._n, self._out_channels,
+                          res.shape[2], res.shape[3])
```
Comment (hnyu) [inline alf/layers.py]: \*res.shape[2:]

File: alf/layers.py [hunk 7]
```diff
@@ -336,6 +336,129 @@ def bias(self):
...
+        return res
+
+    @property
+    def weight(self):
+        # [n*C', C, kernel_size, kernel_size]->[n, C', C, kernel_size, kernel_size]
+        return self._conv2d.weight.view(self._n, self._out_channels,
```
Comment (hnyu) [inline alf/layers.py]: suggest caching this view result with a variable

File: alf/layers.py [hunk 8]
```diff
@@ -404,6 +527,129 @@ def bias(self):
...
+                 padding=0,
+                 use_bias=True,
+                 kernel_initializer=None,
+                 kernel_init_gain=1.0,
+                 bias_init_value=0.0):
+        """A 2D ParallelConvTranspose2D layer that's also responsible for
+        activation and customized weights initialization. An auto gain
+        calculation might depend on the activation following the conv layer.
+        Suggest using this wrapper module instead of ``nn.ConvTranspose2d``
+        if you really care about weight std after init.
```
Comment (hnyu) [inline alf/layers.py]: Add one sentence about the purpose/functionality of this layer

File: alf/layers.py [hunk 9]
```diff
@@ -404,6 +527,129 @@ def bias(self):
...
+                                                   in_channels])
+
+        if use_bias:
+            nn.init.constant_(self._conv_trans2d.bias.data, bias_init_value)
+
+    def forward(self, img):
+        """Forward
+
+        Args:
+            img (torch.Tensor): with shape ``[B, C, H, W]``
```
Comment (hnyu) [inline alf/layers.py:627]: format; explain the difference between with n and without n

File: alf/networks/encoding_networks.py [hunk 10]
```diff
@@ -128,6 +128,81 @@ def forward(self, inputs, state=()):
...
+                    padding=padding))
+            input_channels = filters
+
+    def forward(self, inputs, state=()):
+        """The empty state just keeps the interface same with other networks."""
+        z = inputs
+        for conv_l in self._conv_layers:
+            z = conv_l(z)
+        if self._flatten_output:
+            z = torch.reshape(z, (*z.size()[0:2], -1))
```
Comment (hnyu) [inline alf/networks/encoding_networks.py]: 0:2 -> :2

File: alf/networks/encoding_networks.py [hunk 11]
```diff
@@ -128,6 +128,81 @@ def forward(self, inputs, state=()):
         return z, state
 
 
+@gin.configurable
+class ParallelImageEncodingNetwork(Network):
+    """
+    Parallel Image Encoding Network using CNN.
```
Comment (hnyu) [inline alf/networks/encoding_networks.py]: one sentence explaining its functionality

File: alf/networks/encoding_networks.py [hunk 12]
```diff
@@ -128,6 +128,81 @@ def forward(self, inputs, state=()):
...
+        z = inputs
+        for conv_l in self._conv_layers:
+            z = conv_l(z)
+        if self._flatten_output:
+            z = torch.reshape(z, (*z.size()[0:2], -1))
+        return z, state
+
+
 @gin.configurable
 class ImageDecodingNetwork(Network):
```
Comment (hnyu) [inline alf/networks/encoding_networks.py:224]: No parallel version for ImageDecodingNetwork?

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/layers.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: Added

File: alf/layers.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: changed

File: alf/layers.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: Format updated

File: alf/layers.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: Updated

File: alf/layers.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/layers.py:434]: Updated

File: alf/layers.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: changed

File: alf/layers.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: cached with variable

File: alf/layers.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/layers.py]: added

File: alf/layers.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/layers.py:627]: Format fixed

File: alf/networks/encoding_networks.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/networks/encoding_networks.py]: changed

File: alf/networks/encoding_networks.py (hunk 11 above)
Comment (Haichao-Zhang) [reply alf/networks/encoding_networks.py]: Explanation added

File: alf/networks/encoding_networks.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/networks/encoding_networks.py:224]: ParallelImageDecodingNetwork added

File: alf/layers.py [hunk 13]
```diff
@@ -582,7 +583,7 @@ def __init__(self,
         self._n = n
         self._in_channels = in_channels
         self._out_channels = out_channels
-        self._kernel_size = kernel_size
+        self._kernel_size = common.tuplify2d(kernel_size)
```
Comment (hnyu) [inline alf/layers.py:588]: ParallelConv2D also needs to tuplify2d

File: alf/networks/encoding_networks.py [hunk 14]
```diff
@@ -337,6 +330,126 @@ def forward(self, inputs, state=()):
...
+                    n,
+                    activation=act,
+                    kernel_initializer=kernel_initializer,
+                    strides=strides,
+                    padding=padding))
+            in_channels = filters
+        self._n = n
+
+    def forward(self, inputs, state=()):
+        """Returns an image of shape ``(B,C,H,W)``. The empty state just keeps the
```
Comment (hnyu) [inline alf/networks/encoding_networks.py]: shape comment is incorrect

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/networks/encoding_networks.py (hunk 14 above)
Comment (Haichao-Zhang) [reply alf/networks/encoding_networks.py]: Corrected

File: alf/layers.py (hunk 13 above)
Comment (Haichao-Zhang) [reply alf/layers.py:588]: Added

File: alf/networks/encoding_networks.py (hunk 14 above)
Comment (hnyu) [reply alf/networks/encoding_networks.py]: Inputs can be either one or multiple replicas here?

Comment (Haichao-Zhang) [reply alf/networks/encoding_networks.py]: That's correct. Added more comments and tests on this.

**Review by hnyu — APPROVED** [review]

**Pull request merged**